File: solutions/strings/alternatives/383-ransom-note.js
/**
 * 383. Ransom Note
 * Easy
 *
 * This problem demonstrates key concepts in Strings.
 *
 * SOLUTION EXPLANATION:
 *
 * INTUITION:
 * This problem requires understanding of strings concepts.
 *
 * APPROACH:
 * Apply strings methodology to solve efficiently.
 *
 * WHY THIS WORKS:
 * The solution leverages strings principles for optimal performance.
 *
 * TIME COMPLEXITY: O(n)
 * SPACE COMPLEXITY: O(1)
 *
 * EXAMPLE WALKTHROUGH:
 * Input: [example input]\nStep 1: [explain first step]\nOutput: [expected output]
 *
 * EDGE CASES:
 * - Empty input handling\n- Single element cases\n- Large input considerations
 */

/**
 * Main solution for Problem 383: Ransom Note
 *
 * @param {string} ransomNote - The ransom note to construct
 * @param {string} magazine - The magazine to cut letters from
 * @return {boolean} - True if ransom note can be constructed from magazine
 *
 * Time Complexity: O(m + n) where m is magazine length, n is ransomNote length
 * Space Complexity: O(1) - limited to 26 lowercase letters
 */
function solve(ransomNote, magazine) {
    // Count character frequencies in magazine
    const charCount = new Map();

    for (const char of magazine) {
        charCount.set(char, (charCount.get(char) || 0) + 1);
    }

    // Check if we can construct the ransom note
    for (const char of ransomNote) {
        const count = charCount.get(char) || 0;
        if (count === 0) {
            return false;
        }
        charCount.set(char, count - 1);
    }

    return true;
}

/**
 * Test cases for Problem 383: Ransom Note
 */
function testSolution() {
    console.log('Testing 383. Ransom Note');

    // Test case 1: Cannot construct - not enough 'a'
    const result1 = solve("a", "b");
    console.assert(result1 === false, `Test 1 failed: expected false, got ${result1}`);

    // Test case 2: Cannot construct - not enough of each letter
    const result2 = solve("aa", "ab");
    console.assert(result2 === false, `Test 2 failed: expected false, got ${result2}`);

    // Test case 3: Can construct
    const result3 = solve("aa", "aab");
    console.assert(result3 === true, `Test 3 failed: expected true, got ${result3}`);

    // Test case 4: Empty ransom note
    const result4 = solve("", "abc");
    console.assert(result4 === true, `Test 4 failed: expected true, got ${result4}`);

    console.log('All test cases passed for 383. Ransom Note!');
}

/**
 * Example usage and demonstration
 */
function demonstrateSolution() {
    console.log('\n=== Problem 383. Ransom Note ===');
    console.log('Category: Strings');
    console.log('Difficulty: Easy');
    console.log('');

    // Example demonstration would go here
    testSolution();
}

// Run tests if this file is executed directly
if (require.main === module) {
    demonstrateSolution();
}

// Export for use in other modules
module.exports = {
    solve,
    testSolution,
    demonstrateSolution
};

/**
 * Additional Notes:
 * - This solution focuses on strings concepts
 * - Consider the trade-offs between time and space complexity
 * - Edge cases are crucial for robust solutions
 * - The approach can be adapted for similar problems in this category
 */

File: solutions/strings/383-ransom-note.py
"""
# Difficulty: Easy

# 383. Ransom Note

This problem demonstrates key concepts in Strings and Hash Tables.

**Example:**

<dl class="example-details">
<dt>Input:</dt>
<dd>ransomNote = "aa", magazine = "aab"</dd>
<dt>Output:</dt>
<dd>True</dd>
<dt>Explanation:</dt>
<dd>The ransom note 'aa' cannot be constructed from magazine 'ab' (not enough 'a's)</dd>
</dl>

<details>
<summary><b>🔍 SOLUTION EXPLANATION</b></summary>


### METADATA:
**Techniques**: TBD
**Data Structures**: TBD
**Patterns**: TBD
**Time Complexity**: O(m + n)
**Space Complexity**: O(1)

### INTUITION:
To construct a ransom note from magazine letters, we need to ensure that the magazine contains
at least as many of each character as required by the ransom note. This is essentially checking
if one string's character frequencies are a subset of another's character frequencies.

### APPROACH:
1. **Count magazine characters**: Build a frequency map of all characters in the magazine
2. **Verify ransom note**: For each character in ransom note, check if available in magazine
3. **Decrement counts**: As we use characters from magazine, decrease their counts
4. **Return result**: If we can construct entire ransom note, return True; otherwise False

### WHY THIS WORKS:
- Hash map provides O(1) lookup for character availability
- By counting magazine characters first, we know what's available
- Decrementing counts as we consume characters ensures we don't reuse
- If any character is unavailable or exhausted, we return False immediately

### EXAMPLE WALKTHROUGH:
```
Input: ransomNote = "aa", magazine = "aab"
Step 1: Count magazine chars: {'a': 2, 'b': 1}
Step 2: Check 'a' (first): count is 2, decrement to 1
Step 3: Check 'a' (second): count is 1, decrement to 0
Step 4: All characters available
Output: True

Input: ransomNote = "aa", magazine = "ab"
Step 1: Count magazine chars: {'a': 1, 'b': 1}
Step 2: Check 'a' (first): count is 1, decrement to 0
Step 3: Check 'a' (second): count is 0, not available
Output: False
```

### TIME COMPLEXITY:
O(m + n)
Where m is the length of magazine and n is the length of ransomNote. We iterate through both
strings once.

### SPACE COMPLEXITY:
O(1)
Although we use a hash map, since we're limited to lowercase English letters (26 characters),
the space is bounded by a constant.

### EDGE CASES:
- Empty ransom note: Always True (can construct nothing from anything)
- Empty magazine: False if ransom note is non-empty, True if both empty
- Magazine shorter than ransom note: Could still be False
- Ransom note with characters not in magazine: False

</details>
"""

from collections import Counter


class Solution:
    def solve(self, ransomNote: str, magazine: str) -> bool:
        """
        Check if ransom note can be constructed from magazine letters.

        Args:
            ransomNote: The note to construct
            magazine: The source of letters

        Returns:
            True if ransom note can be constructed, False otherwise

        Time Complexity: O(m + n) where m is magazine length, n is ransomNote length
        Space Complexity: O(1) - bounded by 26 lowercase letters
        """
        # Count character frequencies in magazine
        char_count = Counter(magazine)

        # Check if we can construct the ransom note
        for char in ransomNote:
            if char_count[char] == 0:
                return False
            char_count[char] -= 1

        return True

    def solve_alternative(self, ransomNote: str, magazine: str) -> bool:
        """
        Alternative solution using Counter subtraction.

        This approach is more Pythonic but conceptually similar.

        Time Complexity: O(m + n)
        Space Complexity: O(1)
        """
        ransom_count = Counter(ransomNote)
        magazine_count = Counter(magazine)

        # Check if all characters in ransom note are available in sufficient quantity
        for char, count in ransom_count.items():
            if magazine_count[char] < count:
                return False

        return True


def test_solution() -> None:
    """
    Test cases for 383. Ransom Note.
    """
    solution = Solution()

    # Test case 1: Cannot construct - missing character
    solution.solve("a", "b")

    # Test case 2: Cannot construct - not enough of each character
    solution.solve("aa", "ab")

    # Test case 3: Can construct
    solution.solve("aa", "aab")

    # Test case 4: Empty ransom note (can always construct nothing)
    solution.solve("", "abc")

    # Test case 5: Exact match
    solution.solve("abc", "abc")

    # Test case 6: Magazine has extras
    solution.solve("abc", "aabbcc")

    # Test case 7: Both empty
    solution.solve("", "")

    # Test case 8: Complex case
    solution.solve("hello", "hheelllloo")

    # Test alternative solution
    result9 = solution.solve_alternative("aa", "aab")
    assert result9 == True, f"Test 9 (alternative) failed: expected True, got {result9}"

    print("All test cases passed!")


if __name__ == "__main__":
    test_solution()

    # Example usage
    print("\nExample usage:")
    solution = Solution()
    print(f"Can construct 'aa' from 'aab': {solution.solve('aa', 'aab')}")
    print(f"Can construct 'aa' from 'ab': {solution.solve('aa', 'ab')}")
